src/terms/nonlocal.jl

```diff
@@ -178,22 +178,36 @@ end
 """
 Build form factors (Fourier transforms of projectors) for an atom centered at 0.
 """
-function build_form_factors(psp, qs)
-    qnorms = norm.(qs)
-    T = real(eltype(qnorms))
-    # Compute position-independent form factors
-    form_factors = zeros(Complex{T}, length(qs), count_n_proj(psp))
-    count = 1
-    for l in 0:psp.lmax, m in -l:l
-        prefac_lm = im^l .* ylm_real.(l, m, qs)
-        n_proj_l = size(psp.h[l + 1], 1)
+function build_form_factors(psp, G_plus_ks)
+    T = real(typeof(norm(first(G_plus_ks))))
+
+    # Precompute radial parts at unique |G| and store them in a hash map for O(1) lookup.
+    pq = Dict{T,Matrix{T}}()
+    for Gpk in G_plus_ks
+        q = norm(Gpk)
+        if !haskey(pq, q)
```

Comment:
Great idea! Perhaps (?) an even cleaner way is to use the three-argument form of the `get!` function, i.e.
```
   value = get!(pq, q) do q
      # compute value for q
   end
   # use value
```
Maybe you can also directly inline this with the loop below ?

Reply:
For some reason this adds a bunch of allocation; not sure why.

```julia
# Separate loops
16.045364 seconds (3.51 M allocations: 765.244 MiB, 0.12% gc time)
# get!
16.233814 seconds (11.01 M allocations: 949.460 MiB, 0.10% gc time)
```